Lib/unittest/mock.py

```diff
@@ -3003,6 +3005,98 @@ def __set__(self, obj, val):
         self(val)
 
 
+_timeout_unset = sentinel.TIMEOUT_UNSET
+
+class ThreadingMixin(Base):
+
+    DEFAULT_TIMEOUT = None
+
+    def _get_child_mock(self, /, **kw):
+        if "timeout" in kw:
+            kw["timeout"] = kw.pop("timeout")
+        elif isinstance(kw.get("parent"), ThreadingMixin):
+            kw["timeout"] = kw["parent"]._mock_wait_timeout
+        elif isinstance(kw.get("_new_parent"), ThreadingMixin):
+            kw["timeout"] = kw["_new_parent"]._mock_wait_timeout
+        return super()._get_child_mock(**kw)
+
+    def __init__(self, *args, timeout=_timeout_unset, **kwargs):
+        super().__init__(*args, **kwargs)
+        if timeout is _timeout_unset:
+            timeout = self.DEFAULT_TIMEOUT
+        self.__dict__["_mock_event"] = threading.Event()  # Event for any call
+        self.__dict__["_mock_calls_events"] = []  # Events for each of the calls
+        self.__dict__["_mock_events_lock"] = threading.Lock()
+        self.__dict__["_mock_wait_timeout"] = timeout
+
+    def reset_mock(self, /, *args, **kwargs):
+        """
+        See :func:`.Mock.reset_mock()`
+        """
+        super().reset_mock(*args, **kwargs)
+        self.__dict__["_mock_event"] = threading.Event()
```

Comment:
Wouldn't it make sense to signal existing events before overriding them in case user's code is holding onto them? A condition seems to be necessary to disambiguate call-set events from reset-set events.

Reply:
> Wouldn't it make sense to set existing events before resetting them in case user's code is holding onto them?

But that would invalidate previous `wait` calls that the user has requested isn't it? Why would a user want that?

I might have misunderstood it though, if so, can you write some code that illustrates how this would look like from a user perspective?

Reply:
The invalidation is the point. If a user resets a mock while the test is still waiting on it then it’s a bug IMHO. However, this is an opportunity to handle it gracefully telling the user where they messed up.

I don’t have a legible example in mind, only a misuse.

Reply:
> The invalidation is the point. **If a user resets a mock while the test is still waiting on it** then it’s a bug IMHO. However, this is an opportunity to handle it gracefully telling the user where they messed up.
> 
> I don’t have a legible example in mind, only a misuse.

That would mean that the user test script is multithreaded, one thread that is waiting for the mock to be called and one thread that does the reset_mock.

Must we facilitate multithreaded test scripts?

If so, the methods used to wait for the mock to be called must raise an exception when reset_mock is called.

Reply:
> Must we facilitate multithreaded test scripts?

No, but we should attempt to give a sensible error. It's a testing library, verbosity is expected, at least I'd appreciate it as a user. But by no means it's a blocker.

> If so, the methods used to wait for the mock to be called must raise an exception when reset_mock is called.

Agree.

Reply:
That's interesting. We could add that if desired indeed.
The way I saw `reset_mock` is that it resets the recorded calls, not the expectations I had on the the object.

I'd start with YAGNI TBH as even if we were to implement this I'd suggest to add a kw arg to `reset_mock` to invalidate previous `await_X` calls.